### PR TITLE
Improved Jetpack installation and connection

### LINKED ISSUE #7466: Improve Jetpack Plugin Installation, Setup and Connection
Improve Jetpack plugin installation, setup and connection to .com from the app.

- With a .com oauth token and .org credentials: install, setup and connect Jetpack automagically aka 1 button install.
- Same but only with .org credentials.
- Same but only with an oauth token.

Related issues: https://github.com/wordpress-mobile/WordPress-iOS/issues/5154 https://github.com/wordpress-mobile/WordPress-iOS/issues/5139 https://github.com/wordpress-mobile/WordPress-iOS/issues/4915 https://github.com/wordpress-mobile/WordPress-iOS/issues/3761 https://github.com/wordpress-mobile/WordPress-iOS/issues/7355 https://github.com/wordpress-mobile/WordPress-iOS/issues/7339

### PR DESCRIPTION
**Pull request by koke**
This new system allows for easier installation, activation, and connection of Jetpack from the app itself.

Previously, we'd throw the user into a wp-admin web view and let them figure out what to do. With this new implementation, we use the wizard at http://wordpress.com/jetpack/connect to help guide the users through the process.

If the user is not already authenticated to WordPress.com, it will use the native login flow to ask for credentials. Currently there's no option to sign up from there so an existing account is required, but @nheagy mentioned he's working on sign up, so it shouldn't be too hard to support that once that's ready.

Adding also @nheagy as a reviewer to verify the changes to login code are 👌 

Fixes #7466

To test:

Add a self hosted site without Jetpack and try to install and connect from the apps.
The site could have Jetpack not installed, installed but not active, or active but not connected.
Also, login could use magic login, password, or Google sign in. I think I've tested all the combinations, but I'd love some extra help testing this.

### CODE REVIEW

**Comment by elibud** [comment]
@nheagy do you want to split the testing?

- [ ] Login with username + password + 2FA
- [ ] Login with magic link
- [ ] Login with Google
- [ ] Jetpack not installed
- [ ] Jetpack installed but not active
- [ ] Jetpack active but not connected

**Comment by elibud** [comment]
For a site with Jetpack installed but not activated we get this screen: 

simulator screen shot - iphone 8 - 2018-01-16 at 11 06 35

Which is not really accurate, but I guess is not a big deal (I doubt users keep in mind whether it's installed, installed but not active, etc), @koke what do you think?

@seear can we get rid of the top bar for the mobile flow?

simulator screen shot - iphone 8 - 2018-01-16 at 11 08 36

**Comment by koke** [comment]
Yes, that "install" isn't accurate, but not sure if we can detect better. I kind of left it because there's a "Improve Jetpack connection status detection" on the remaining tasks list and I was going to look into it as part of that. In any case, we might want to tweak it to something more generic like "Set up Jetpack"

**Comment by elibud** [comment]
For Jetpack installed but not activated there are two issues, I guess both for you @seear 

1. The call to action to activate Jetpack is not in the Activate task but after Connect
2. I'm suddenly taken to `wp-admin` after I tap activate, is this expected?

Both can be seen below:

jetpackinstalledbutnotactivated

**Comment by elibud** [comment]
> "Set up Jetpack"

☝️ yeah that does the trick

**Comment by nheagy** [comment]
@elibud sounds good for testing. I'll test the last two

**Comment by koke** [comment]
I guess I've been looking at the flow to much to notice 😁 

> The call to action to activate Jetpack is not in the Activate task but after Connect

If you tap on the image it also works, but yeah it's not so obvious

> I'm suddenly taken to wp-admin after I tap activate, is this expected?

Yep, that isn't great but I think it's the best we can do. That's the page were you activate the plugin, there's just a lot of other stuff around the "activate" link. Once this is in place we could start experimenting with injecting JS/CSS to make things more obvious or even click on behalf of the user.

File: WordPress/Classes/ViewRelated/Jetpack/JetpackConnectionWebViewController.swift [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+    func isSiteLogin(url: URL) -> Bool {
+        guard let loginURL = URL(string: blog.loginUrl()) else {
+            return false
+        }
+
+        return url.matchesPath(in: loginURL)
+    }
+
+    func isSiteAdmin(path: String) -> (URL) -> Bool {
```
Comment (elibud) [inline WordPress/Classes/ViewRelated/Jetpack/JetpackConnectionWebViewController.swift]: It took me a while to figure out this one.

File: WordPress/Classes/ViewRelated/Jetpack/JetpackConnectionWebViewController.swift [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        webView.load(request)
+    }
+
+    func presentDotComLogin(redirect: URL) {
+        pendingDotComRedirect = redirect
+        observeLoginNotifications(true)
+        SigninHelpers.showLoginForJustWPComFromPresenter(self, forJetpackBlog: blog)
+    }
+
+    func observeLoginNotifications(_ observe: Bool) {
```
Comment (elibud) [inline WordPress/Classes/ViewRelated/Jetpack/JetpackConnectionWebViewController.swift]: I'm not crazy about this, I would rather have two methods `startObservingLoginNotifications` and `stopObservingLoginNotifications`. I can see this is similar to what we have in `JetpackLoginViewController` and I like it less there with the return in between, but I'm probably just being nipticky, feel free go completely ignore me on this one.

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift [hunk 3]
```diff
@@ -156,12 +156,13 @@ import WordPressShared
         // NUX vc then present the auth controller.
         // - If the rootViewController is presenting *any* other vc, present the
         // auth controller from the presented vc.
-        if let presenter = rootViewController.presentedViewController, presenter.isKind(of: NUXNavigationController.self) || presenter.isKind(of: LoginNavigationController.self) {
-            rootViewController.dismiss(animated: false, completion: {
-                rootViewController.present(navController, animated: false, completion: nil)
+        let presenter = controllerForAuthControllerPresenter(rootViewController)
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: This looks great 👍

**Comment by seear** [comment]
> can we get rid of the top bar for the mobile flow?

Yep, that will happen when [this PR](https://github.com/Automattic/wp-calypso/pull/21408) lands.

> I'm suddenly taken to wp-admin after I tap activate, is this expected?

The only way to activate the Jetpack plugin is to hit _Activate_ in wp-admin, so this is necessary.

> The call to action to activate Jetpack is not in the Activate task but after Connect

I think the idea there is to help the reader understand that they need to hit _Connect_ in wp-admin after hitting _Activate_.

One of the next projects up for Luna is Jetpack auto-install, where user supplies creds and we install and activate plugin on their behalf, which should alleviate some of these issues.

**Comment by koke** [comment]
@elibud ready for another look, thanks 🙇

**Comment by elibud** [comment]
> One of the next projects up for Luna is Jetpack auto-install, where user supplies creds and we install and activate plugin on their behalf, which should alleviate some of these issues.

@seear please make sure to keep us in the loop when that happens :)

**Comment by elibud** [comment]
@koke the code looks great but I've seen to have entered a loop with, ironically, the site http://justforyoukoke.mystagingwebsite.com 

I add it as a self-hosted site, then go to Stats, this is what I see:

simulator screen shot - iphone 8 - 2018-01-18 at 11 41 03

I follow the steps and once the Login VC is dismissed I end up in the same ☝️ screen again. I'll send you credentials so you can test it.

**Comment by elibud** [comment]
@koke let me know if this is ready for another pass.

**Comment by koke** [comment]
Almost there, waiting for this one to be merged for some extra testing https://github.com/Automattic/wp-calypso/pull/21729

**Comment by koke** [comment]
Added localization. Finally ready for another look 😂

**Comment by koke** [comment]
giphy

**Pull request merged**